java-checks/src/main/java/org/sonar/java/checks/StringCallsBeyondBoundsCheck.java

```diff
@@ -0,0 +1,154 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+
+import org.sonar.check.Rule;
+import org.sonar.java.checks.helpers.ExpressionsHelper;
+import org.sonar.java.checks.methods.AbstractMethodDetection;
+import org.sonar.java.matcher.MethodMatcher;
+import org.sonar.plugins.java.api.tree.Arguments;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.Tree.Kind;
+
+@Rule(key = "S3039")
+public class StringCallsBeyondBoundsCheck extends AbstractMethodDetection {
+
+  private static final String STRING = "java.lang.String";
+  private static final MethodMatcher STRING_LENGTH =
+    MethodMatcher.create().typeDefinition(STRING).name("length").withoutParameter();
+
+  private static interface Condition {
+    boolean apply(ExpressionTree string, Arguments arguments);
+  }
+
+  @Override
+  protected List<MethodMatcher> getMethodInvocationMatchers() {
+    return Arrays.asList(
+      MethodMatcher.create().typeDefinition(STRING).name("charAt").addParameter("int"),
+      MethodMatcher.create().typeDefinition(STRING).name("codePointAt").addParameter("int"),
+      MethodMatcher.create().typeDefinition(STRING).name("codePointBefore").addParameter("int"),
+      MethodMatcher.create().typeDefinition(STRING).name("codePointCount").addParameter("int").addParameter("int"),
+      MethodMatcher.create().typeDefinition(STRING).name("getChars").addParameter("int").addParameter("int").addParameter("char[]").addParameter("int"),
+      MethodMatcher.create().typeDefinition(STRING).name("offsetByCodePoints").addParameter("int").addParameter("int"),
+      MethodMatcher.create().typeDefinition(STRING).name("substring").addParameter("int"),
+      MethodMatcher.create().typeDefinition(STRING).name("substring").addParameter("int").addParameter("int"),
+      MethodMatcher.create().typeDefinition(STRING).name("subSequence").addParameter("int").addParameter("int")
+    );
+  }
+
+  @Override
+  protected void onMethodInvocationFound(MethodInvocationTree invocation) {
+    boolean issue = false;
+    String method = invocation.symbol().name();
+    switch (method) {
+      case "charAt":
+      case "codePointAt":
+        // we cannot use `!lt` instead of `gte` because it would evaluate to true on an empty optional
+        issue = check(invocation, (str, args) -> lt(val(args.get(0)), opt(0)) || gte(val(args.get(0)), len(str)));
```

Comment:
these conditions are a bit hard to reason about, I believe it's because they work on too low abstraction level. It would be more readable if you defined functions on higher abstraction. You can have a function like `isValidIndex(ExpressionTree indexParam, int lowerBound, int upperBound)`,  then for example `isValidStringIndex(ExpressionTree str, ExpressionTree index)` which calls the previous with `0, len(str)`.... Once you have such functions you will be able to simplify many of the expressions, and more importantly, it will be easier to reason about conditions.